File: sections/24_grid_dp/problems/c_limited_turns/solution.py
import sys


MOD = 1_000_000_007
RIGHT = 0
DOWN = 1


def main() -> None:
    input = sys.stdin.readline
    first = input().split()
    if not first:
        return
    n, m, k = map(int, first)
    grid = [input().strip() for _ in range(n)]
    if grid[0][0] == "#" or grid[n - 1][m - 1] == "#":
        print(0)
        return
    dp = [[[[0, 0] for _ in range(k + 1)] for _ in range(m)] for _ in range(n)]
    if m > 1 and grid[0][1] == ".":
        dp[0][1][0][RIGHT] = 1
    if n > 1 and grid[1][0] == ".":
        dp[1][0][0][DOWN] = 1
    if n == 1 and m == 1:
        print(1)
        return
    for r in range(n):
        for c in range(m):
            if grid[r][c] == "#":
                continue
            for turns in range(k + 1):
                if c > 0 and grid[r][c - 1] == ".":
                    dp[r][c][turns][RIGHT] = (
                        dp[r][c][turns][RIGHT] + dp[r][c - 1][turns][RIGHT]
                    ) % MOD
                    if turns:
                        dp[r][c][turns][RIGHT] = (
                            dp[r][c][turns][RIGHT] + dp[r][c - 1][turns - 1][DOWN]
                        ) % MOD
                if r > 0 and grid[r - 1][c] == ".":
                    dp[r][c][turns][DOWN] = (
                        dp[r][c][turns][DOWN] + dp[r - 1][c][turns][DOWN]
                    ) % MOD
                    if turns:
                        dp[r][c][turns][DOWN] = (
                            dp[r][c][turns][DOWN] + dp[r - 1][c][turns - 1][RIGHT]
                        ) % MOD
    ans = 0
    for turns in range(k + 1):
        ans = (ans + dp[n - 1][m - 1][turns][RIGHT] + dp[n - 1][m - 1][turns][DOWN]) % MOD
    print(ans)


if __name__ == "__main__":
    main()

File: sections/24_grid_dp/problems/c_limited_turns/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    const int MOD = 1000000007;
    const int RIGHT = 0, DOWN = 1;
    int n, m, K;
    if (!(cin >> n >> m >> K)) return 0;
    vector<string> grid(n);
    for (string &row : grid) cin >> row;
    if (grid[0][0] == '#' || grid[n - 1][m - 1] == '#') {
        cout << 0 << '\n';
        return 0;
    }
    if (n == 1 && m == 1) {
        cout << 1 << '\n';
        return 0;
    }
    vector dp(n, vector(m, vector(K + 1, array<int, 2>{0, 0})));
    if (m > 1 && grid[0][1] == '.') dp[0][1][0][RIGHT] = 1;
    if (n > 1 && grid[1][0] == '.') dp[1][0][0][DOWN] = 1;
    for (int r = 0; r < n; r++) {
        for (int c = 0; c < m; c++) {
            if (grid[r][c] == '#') continue;
            for (int turns = 0; turns <= K; turns++) {
                if (c > 0 && grid[r][c - 1] == '.') {
                    int val = dp[r][c][turns][RIGHT];
                    val += dp[r][c - 1][turns][RIGHT];
                    if (val >= MOD) val -= MOD;
                    if (turns) {
                        val += dp[r][c - 1][turns - 1][DOWN];
                        if (val >= MOD) val -= MOD;
                    }
                    dp[r][c][turns][RIGHT] = val;
                }
                if (r > 0 && grid[r - 1][c] == '.') {
                    int val = dp[r][c][turns][DOWN];
                    val += dp[r - 1][c][turns][DOWN];
                    if (val >= MOD) val -= MOD;
                    if (turns) {
                        val += dp[r - 1][c][turns - 1][RIGHT];
                        if (val >= MOD) val -= MOD;
                    }
                    dp[r][c][turns][DOWN] = val;
                }
            }
        }
    }
    int ans = 0;
    for (int turns = 0; turns <= K; turns++) {
        ans += dp[n - 1][m - 1][turns][RIGHT];
        if (ans >= MOD) ans -= MOD;
        ans += dp[n - 1][m - 1][turns][DOWN];
        if (ans >= MOD) ans -= MOD;
    }
    cout << ans << '\n';
    return 0;
}

